### ISSUE TITLE
Rule Proposal: Catch no await inside a catch block

### ISSUE THREAD

**Issue by erlichmen** [eslint/eslint#13884]
consider the following code:

```typescript
async function foo(val): number {
  if (!val) {
     throw new Error();
   }

  return 42;
}

function bar1(): number {
  try {
     return foo(0);
  } catch (err) {
     return -1;
  }
}

function bar2(): number {
  try {
     return await foo(0);
  } catch (err) {
     return -1;
  }
}

```
the function bar1() is probably a bug as it does not call await and the catch block will not be handled in the context of the function as intended while bar2() function is the current implementation.

eslint should probably warn about a potential error when evaluating bar1()

**Why should this rule be included in ESLint (instead of a plugin)?**
This seems to be in the code error that every js developer might stumble upon.

**Are you willing to submit a pull request to implement this rule?**
With proper guidance sure.

**Comment by yeonjuan** [comment]
@erlichmen Hi:) Thanks for the proposal.

IMO, It cannot be fully handled in eslint, because we cannot know whether `foo()` returns `promise` or not. (And also don't know about throwing error in `foo()`). Of course, we can analyze it if they are in the same file but we cannot do that for the imported function or 3rd party library.

**Comment by erlichmen** [comment]
I was under the impression that no-floating-promises already does half the
work as it knows that the promise was not handled properly.

**Comment by mdjermanovic** [comment]
> I was under the impression that no-floating-promises already does half the work as it knows that the promise was not handled properly.

`no-floating-promises` is a `typescript-eslint` rule: [typescript-eslint/no-floating-promises](https://github.com/typescript-eslint/typescript-eslint/blob/master/packages/eslint-plugin/docs/rules/no-floating-promises.md).

We could check if the called function is `async` (and thus non-throwing), but that would be limited to functions in the same file so we avoid adding such rules to the core.

**Comment by ljharb** [comment]
Why is this "probably a bug"? As written, it suggests that sync errors are intentionally caught and async errors are intentionally returned to the caller as a rejection.

**Comment by mdjermanovic** [comment]
Also, doesn't typescript already report this as returning `Promise<number>` instead of `number`?

> ```ts
> function bar1(): number {
>   try {
>      return foo(0); // <-- error?
>   } catch (err) {
>      return -1;
>   }
> }
> ```

**Comment by ljharb** [comment]
Presumably it's meant to be `async function`, otherwise `await` wouldn't be valid in it anyways?

**Comment by erlichmen** [comment]
My bad, I'll go to eslint-typescript and bother them, I don't know if this rule can be of any use in the eslint world if so keep this  issue otherwise it can be closed and yes I forgot to but the async there in bar()
